File: GIMMECore/source/Adaptation/AlgDefStructs/RegressionAlg.cpp
#include "../../../headers/Adaptation/AlgDefStructs/RegressionAlg.h"

//---------------------- KNNRegression stuff ---------------------------
KNNRegression::KNNRegression(int numberOfNNs)
{
	this->numberOfNNs = numberOfNNs;
}

PlayerState KNNRegression::predict(InteractionsProfile profile, Player* player)
{
	std::vector<PlayerState> pastModelIncs = player->getPastModelIncreases();
	std::vector<PlayerState> pastModelncsCopy = std::vector<PlayerState>(pastModelIncs);
	int pastModelIncsSize = (int)pastModelIncs.size();

	PlayerState predictedModel = { profile, 0 , 0 };
	std::sort(pastModelncsCopy.begin(), pastModelncsCopy.end(), InteractionsProfileSort(profile));

	if (pastModelIncsSize > numberOfNNs) {
		pastModelncsCopy.resize(numberOfNNs);
	}
	int pastModelncsCopySize = (int)pastModelncsCopy.size();

	for (int i = 0; i < pastModelncsCopySize; i++) {
		InteractionsProfile pastProfile = pastModelncsCopy[i].profile;
		double distance = profile.distanceBetween(pastProfile);

		predictedModel.characteristics.ability += (pastModelncsCopy[i].characteristics.ability* (1 - distance)) / (double)(pastModelncsCopySize); //* (1 - distance) 
		predictedModel.characteristics.engagement += (pastModelncsCopy[i].characteristics.engagement* (1 - distance)) / (double)(pastModelncsCopySize); //* (1 - distance)
	}

	return predictedModel;
}



//---------------------- NeuralNetworkRegression stuff ---------------------------

NeuralNetworkRegression::NeuralNetworkRegression()
{
}

PlayerState NeuralNetworkRegression::predict(InteractionsProfile profile, Player* player)
{
	PlayerState predictedModel = { profile, 0 , 0 };

	return predictedModel;
}



//---------------------- ReinforcementLearningRegression stuff ---------------------------

ReinforcementLearningRegression::ReinforcementLearningRegression()
{
}

PlayerState ReinforcementLearningRegression::predict(InteractionsProfile profile, Player* player)
{
	PlayerState predictedModel = { profile, 0 , 0 };

	return predictedModel;
}

File: GIMMECore/Adaptation/AlgDefStructs/RegressionAlg.py
from abc import ABC, abstractmethod


class RegressionAlg(ABC):

	@abstractmethod
	def predict(profile, player):
		pass

# ---------------------- KNNRegression stuff ---------------------------
class KNNRegression(RegressionAlg):

	def __init__(self, numberOfNNs):
		self.numberOfNNs = numberOfNNs;

	def interactionsProfileSort(pState1, pState2):

		dist1 = testedProfile.distanceBetween(pState1.profile)
		dist2 = testedProfile.distanceBetween(pState2.profile)

		pState1.dist = dist1
		pState2.dist = dist2

		return dist1 < dist2

	def predict(profile, player):
		pastModelIncs = player.getPastModelIncreases()
		pastModelIncsCopy = pastModelIncs.copy()
		pastModelIncsSize = len(pastModelIncs)

		predictedModel = PlayerState( profile, PlayerCharacteristics() , 0 )
		predictedModel = sorted(predictedModel, key=InteractionsProfileSort(profile))

		if (pastModelIncsSize > numberOfNNs):
			pastModelIncsCopy = numpy.resize(pastModelIncsCopy, numberOfNNs)
	
		pastModelIncsCopySize = len(pastModelncsCopy)

		for i in range(pastModelncsCopySize):
			pastProfile = pastModelIncsCopy[i].profile;
			distance = profile.distanceBetween(pastProfile);

			predictedModel.characteristics.ability += (pastModelncsCopy[i].characteristics.ability* (1 - distance)) / (double)(pastModelncsCopySize); #* (1 - distance) 
			predictedModel.characteristics.engagement += (pastModelncsCopy[i].characteristics.engagement* (1 - distance)) / (double)(pastModelncsCopySize); #* (1 - distance)
		
		return predictedModel;


# ---------------------- NeuralNetworkRegression stuff ---------------------------
class NeuralNetworkRegression(RegressionAlg):

	def predict(profile, player):
		pass


# ---------------------- ReinforcementLearningRegression stuff ---------------------------
class ReinforcementLearningRegression(RegressionAlg):

	def predict(profile, player):
		pass

